### PR TITLE
[msan] Implement shadow propagation for _mm_dp_pd, _mm_dp_ps, _mm256_dp_ps

### PR DESCRIPTION
**Pull request by vitalybuka**
Default intrinsic handling was to report any
uninitialized part of argument. However intrinsics
use mask which allow to ignore parts of input, so
it's OK to have vectors partially initialized.

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 1]
```diff
@@ -3287,6 +3287,76 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
     setOriginForNaryOp(I);
   }
 
+  // Convert `Mask` into `<n x i1>`.
+  Constant *createDppMask(unsigned Width, unsigned Mask) {
+    SmallVector<Constant *, 4> R;
+    R.assign(Width, ConstantInt::getFalse(F.getContext()));
+    for (auto &M : R) {
+      if (Mask & 1)
+        M = ConstantInt::getTrue(F.getContext());
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:3296]: wouldn't this bge more readable as

```
M = ConstantInt::getBool(F.getContext(), Mask & 1)
```

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 2]
```diff
@@ -3287,6 +3287,76 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
...
+      Mask >>= 1;
+    }
+    return ConstantVector::get(R);
+  }
+
+  // Calculate output shadow as array of booleans `<n x i1>`, assuming if any
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:3302]: "assuming" is a bit weird. Isn't that how the poison should propagate, not an assumption?

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 3]
```diff
@@ -3287,6 +3287,76 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
...
+  // arg is poisoned, entire dot product is poisoned.
+  Value *makeDppShadowI1(IRBuilder<> &IRB, Value *S, unsigned SrcMask,
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:3304]: What does `I1` stand for?

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 2 above)
Comment (vitalybuka) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:3302]: Not exactly. Input bits can affect particular output bits.
E.g. argument with single uninitialized bit can be multiplied by initialized zero.
So it's approximation, but better checking will be expensive.

Comment (fmayer) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:3302]: Ah, ok. Please add a comment. I guess these instructions are beyond the reach of the C++ standard, so there is no proper definition for "undefined bits" that would say whether multiplying one by zero is UB or not?

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 3 above)
Comment (vitalybuka) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:3304]: `i1` type of shadow elements

renamed the function

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp (hunk 2 above)
Comment (vitalybuka) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:3300]: unfortunately have have to do such shortcuts even for C++ generated code

Comment (fmayer) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:3300]: OK

**Review by fmayer — APPROVED** [review]

**Pull request merged**